pkg/k8s/watchers/network_policy.go

```diff
@@ -4,76 +4,65 @@
 	"github.com/cilium/cilium/pkg/source"
 )
 
-func (k *K8sWatcher) networkPoliciesInit(slimClient slimclientset.Interface, swgKNPs *lock.StoppableWaitGroup) {
-	apiGroup := k8sAPIGroupNetworkingV1Core
-	store, policyController := informer.NewInformer(
-		utils.ListerWatcherFromTyped[*slim_networkingv1.NetworkPolicyList](
-			slimClient.NetworkingV1().NetworkPolicies("")),
-		&slim_networkingv1.NetworkPolicy{},
-		0,
-		cache.ResourceEventHandlerFuncs{
-			AddFunc: func(obj interface{}) {
-				var valid, equal bool
-				defer func() { k.K8sEventReceived(apiGroup, metricKNP, resources.MetricCreate, valid, equal) }()
-				if k8sNP := k8s.ObjToV1NetworkPolicy(obj); k8sNP != nil {
-					valid = true
-					err := k.addK8sNetworkPolicyV1(k8sNP)
-					k.K8sEventProcessed(metricKNP, resources.MetricCreate, err == nil)
-				}
-			},
-			UpdateFunc: func(oldObj, newObj interface{}) {
-				var valid, equal bool
-				defer func() { k.K8sEventReceived(apiGroup, metricKNP, resources.MetricUpdate, valid, equal) }()
-				if oldK8sNP := k8s.ObjToV1NetworkPolicy(oldObj); oldK8sNP != nil {
-					if newK8sNP := k8s.ObjToV1NetworkPolicy(newObj); newK8sNP != nil {
-						valid = true
-						if oldK8sNP.DeepEqual(newK8sNP) {
-							equal = true
-							return
-						}
-
-						err := k.updateK8sNetworkPolicyV1(oldK8sNP, newK8sNP)
-						k.K8sEventProcessed(metricKNP, resources.MetricUpdate, err == nil)
-					}
-				}
-			},
-			DeleteFunc: func(obj interface{}) {
-				var valid, equal bool
-				defer func() { k.K8sEventReceived(apiGroup, metricKNP, resources.MetricDelete, valid, equal) }()
-				k8sNP := k8s.ObjToV1NetworkPolicy(obj)
-				if k8sNP == nil {
-					return
-				}
-
-				valid = true
-				err := k.deleteK8sNetworkPolicyV1(k8sNP)
-				k.K8sEventProcessed(metricKNP, resources.MetricDelete, err == nil)
-			},
-		},
-		nil,
-	)
-	k.networkpolicyStore = store
-	k.blockWaitGroupToSyncResources(k.stop, swgKNPs, policyController.HasSynced, k8sAPIGroupNetworkingV1Core)
-	go policyController.Run(k.stop)
+func (k *K8sWatcher) NetworkPoliciesInit() {
+	k.networkPoliciesInitOnce.Do(func() {
+		var synced atomic.Bool
+		swg := lock.NewStoppableWaitGroup()
+		k.blockWaitGroupToSyncResources(k.stop, swg, func() bool { return synced.Load() }, k8sAPIGroupNetworkingV1Core)
+		go k.networkPolicyEventLoop(&synced)
+		swg.Wait()
+		k.k8sAPIGroups.AddAPI(k8sAPIGroupNetworkingV1Core)
+	})
+}
 
-	k.k8sAPIGroups.AddAPI(apiGroup)
+func (k *K8sWatcher) networkPolicyEventLoop(synced *atomic.Bool) {
+	apiGroup := k8sAPIGroupNetworkingV1Core
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	events := k.resources.NetworkPolicies.Events(ctx)
+	for {
+		select {
+		case <-k.stop:
+			cancel()
+		case event, ok := <-events:
+			if !ok {
+				return
+			}
+			var err error
+			switch event.Kind {
+			case resource.Sync:
+				var store resource.Store[*slim_networkingv1.NetworkPolicy]
+				synced.Store(true)
+				store, err = k.resources.NetworkPolicies.Store(ctx)
```

Comment:
is it safe to close storeSet if err is non-nil here?

Reply:
ugh, no that's an oversight. Thanks!